documentation/testing_plans.md

```diff
@@ -0,0 +1,808 @@
+| [`upload_file`](plan_functions.md#upload-file) | [`allow_upload`](boltspec_reference.md#allow-upload) | [`expect_upload`](boltspec_reference.md#expect-upload) |
+
+### Modifiers
+
+Stubs and mocks support modifiers that allow you to add specificity to your
+tests. Modifiers do any of the following:
+
+- Restrict the number of times an executor function can be invoked or is
+  expected to be invoked.
+
+- Specify the arguments that an executor function is invoked with.
+
+- Set the value returned by an executor function.
+
+For example, you can use the `be_called_times` modifier to limit the number of
+times you expect a task to be run in a plan. The following test sets an
+assertion that the `facts` task is only called once:
+
+```ruby
+it 'does not run more than two tasks' do
+  # Assert that the 'facts' task is run exactly once
+  expect_task('facts').be_called_times(1)
+
+  # Run the plan
+  run_plan('my_project::my_plan', 'servers' => 'servers', 'databases' => 'databases')
+end
+```
+
+This test would pass for the following plan, since the plan only runs the
+`facts` task once:
+
+```puppet
+plan my_project::my_plan (
+  TargetSpec $servers,
+  TargetSpec $databases
+) {
+  $targets = $servers + $databases
+  $results = run_task('facts', $targets)
+  return $results
+}
+```
+
+However, the test would fail for the following plan, since the plan runs the
+`facts` task twice:
+
+```puppet
+plan my_project::my_plan (
+  TargetSpec $servers,
+  TargetSpec $databases
+) {
+  $server_results   = run_task('facts', $servers)
+  $database_results = run_task('facts', $databases)
+  $results          = ResultSet.new($server_results.results + $database_results.results)
+  return $results
+}
+```
+
+Modifiers can also be chained, allowing you to add as much or as little
+specificity to a stub or mock as you want. For example, you can restrict a task
+to run on a specific list of targets with a specific set of parameters:
+
+```ruby
+it 'configures servers' do
+  # Assert that the 'my_project::configure' task is run
+  expect_task('my_project::configure').with_targets('servers').with_params({ 'confpath' => '/path/to/conf' })
+
+  # Run the plan
+  run_plan('my_project::configure', 'targets' => ['servers', 'databases'])
+end
+```
+
+📖  **Related information**
+
+- See the [BoltSpec reference](boltspec_reference.md) for a full list of
+  modifiers.
+
+### Matching stubs and mocks
+
+When you run your tests and a plan invokes an executor function, `BoltSpec`
+matches the function invocation to any stubs or mocks created for the test.
+Executor function invocations are matched to any stub or mock that is as
+specific or less specific than the function invocation, but not to any stub or
+mock that is more specific than the function invocation. If the function
+invocation matches multiple stubs or mocks, it uses the last stub or mock that
+matched.
```

Comment:
```suggestion
When you run your tests and a plan invokes an executor function, `BoltSpec`
matches the function invocation to any stubs or mocks created for the test.
`BoltSpec` matches executor function invocations to any stub or mock that is as
specific or less specific than the function invocation, but not to any stub or
mock that is more specific than the function invocation. If the function
invocation matches multiple stubs or mocks, `BoltSpec` uses the last stub or mock 
that matched.

For a test to pass, `BoltSpec` must find a match for each stub or mock in the test. 
```
Adding that last bit because I think it's what I got tripped up on later.

Reply:
👍 Added. Modified it to only say 'each mock', since stubs don't make assertions and thus don't need to be matched for a test to pass.